File: pkg/transport/jsonrpc.go
package transport

import (
	"encoding/json"
	"fmt"
)

// JSONRPCMessage represents a JSON-RPC message
type JSONRPCMessage struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      interface{}     `json:"id,omitempty"`
	Method  string          `json:"method,omitempty"`
	Params  json.RawMessage `json:"params,omitempty"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *JSONRPCError   `json:"error,omitempty"`
}

// JSONRPCError represents a JSON-RPC error
type JSONRPCError struct {
	Code    int             `json:"code"`
	Message string          `json:"message"`
	Data    json.RawMessage `json:"data,omitempty"`
}

// NewRequestMessage creates a new JSON-RPC request message
func NewRequestMessage(method string, params interface{}, id interface{}) (*JSONRPCMessage, error) {
	var paramsJSON json.RawMessage
	if params != nil {
		var err error
		paramsJSON, err = json.Marshal(params)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal params: %w", err)
		}
	}

	return &JSONRPCMessage{
		JSONRPC: "2.0",
		Method:  method,
		Params:  paramsJSON,
		ID:      id,
	}, nil
}

// NewResponseMessage creates a new JSON-RPC response message
func NewResponseMessage(id interface{}, result interface{}) (*JSONRPCMessage, error) {
	var resultJSON json.RawMessage
	if result != nil {
		var err error
		resultJSON, err = json.Marshal(result)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal result: %w", err)
		}
	}

	return &JSONRPCMessage{
		JSONRPC: "2.0",
		Result:  resultJSON,
		ID:      id,
	}, nil
}

// NewErrorMessage creates a new JSON-RPC error message
func NewErrorMessage(id interface{}, code int, message string, data interface{}) (*JSONRPCMessage, error) {
	var dataJSON json.RawMessage
	if data != nil {
		var err error
		dataJSON, err = json.Marshal(data)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal error data: %w", err)
		}
	}

	return &JSONRPCMessage{
		JSONRPC: "2.0",
		Error: &JSONRPCError{
			Code:    code,
			Message: message,
			Data:    dataJSON,
		},
		ID: id,
	}, nil
}

// NewNotificationMessage creates a new JSON-RPC notification message
func NewNotificationMessage(method string, params interface{}) (*JSONRPCMessage, error) {
	var paramsJSON json.RawMessage
	if params != nil {
		var err error
		paramsJSON, err = json.Marshal(params)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal params: %w", err)
		}
	}

	return &JSONRPCMessage{
		JSONRPC: "2.0",
		Method:  method,
		Params:  paramsJSON,
	}, nil
}

// IsRequest returns true if the message is a request
func (m *JSONRPCMessage) IsRequest() bool {
	return m.Method != "" && m.ID != nil
}

// IsResponse returns true if the message is a response
func (m *JSONRPCMessage) IsResponse() bool {
	return m.ID != nil && (m.Result != nil || m.Error != nil) && m.Method == ""
}

// IsNotification returns true if the message is a notification
func (m *JSONRPCMessage) IsNotification() bool {
	return m.Method != "" && m.ID == nil
}

// Validate validates the JSON-RPC message
func (m *JSONRPCMessage) Validate() error {
	if m.JSONRPC != "2.0" {
		return fmt.Errorf("invalid JSON-RPC version: %s", m.JSONRPC)
	}

	if !m.IsRequest() && !m.IsResponse() && !m.IsNotification() {
		return fmt.Errorf("invalid JSON-RPC message format")
	}

	return nil
}

File: src/transport/jsonrpc.rs
use serde::{Deserialize, Serialize};

/// JSON-RPC error structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JsonRpcError {
    pub code: i32,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
}

/// JSON-RPC message structure that follows the MCP protocol specification
/// This can represent any of the three message types defined in the MCP spec:
/// 1. Requests: { jsonrpc: "2.0", id: string|number, method: string, params?: object }
/// 2. Responses: { jsonrpc: "2.0", id: string|number, result?: object, error?: { code: number, message: string, data?: any } }
/// 3. Notifications: { jsonrpc: "2.0", method: string, params?: object }
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JsonRpcMessage {
    pub jsonrpc: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub method: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<JsonRpcError>,
}

impl JsonRpcMessage {
    /// Create a new request message
    pub fn new_request(
        method: &str,
        params: Option<serde_json::Value>,
        id: serde_json::Value,
    ) -> Self {
        Self {
            jsonrpc: "2.0".to_string(),
            id: Some(id),
            method: Some(method.to_string()),
            params,
            result: None,
            error: None,
        }
    }

    /// Create a new response message
    pub fn new_response(id: serde_json::Value, result: serde_json::Value) -> Self {
        Self {
            jsonrpc: "2.0".to_string(),
            id: Some(id),
            method: None,
            params: None,
            result: Some(result),
            error: None,
        }
    }

    /// Create a new error response message
    pub fn new_error(
        id: serde_json::Value,
        code: i32,
        message: &str,
        data: Option<serde_json::Value>,
    ) -> Self {
        Self {
            jsonrpc: "2.0".to_string(),
            id: Some(id),
            method: None,
            params: None,
            result: None,
            error: Some(JsonRpcError {
                code,
                message: message.to_string(),
                data,
            }),
        }
    }

    /// Create a new notification message
    pub fn new_notification(method: &str, params: Option<serde_json::Value>) -> Self {
        Self {
            jsonrpc: "2.0".to_string(),
            id: None,
            method: Some(method.to_string()),
            params,
            result: None,
            error: None,
        }
    }

    /// Check if this is a request message
    pub fn is_request(&self) -> bool {
        self.id.is_some() && self.method.is_some()
    }

    /// Check if this is a response message
    pub fn is_response(&self) -> bool {
        self.id.is_some() && (self.result.is_some() || self.error.is_some())
    }

    /// Check if this is a notification message
    pub fn is_notification(&self) -> bool {
        self.id.is_none() && self.method.is_some()
    }
}